File: src/aoc2023/day3b.rs
use std::collections::HashMap;
use std::io::BufRead;

pub fn run() {
    let mut numbers: Vec<Vec<(u32, usize, usize)>> = Vec::new();
    let mut symbols: Vec<Vec<usize>> = Vec::new();
    let mut gears: HashMap<(usize, usize), Vec<u32>> = HashMap::new();

    std::io::stdin()
        .lock()
        .lines()
        .enumerate()
        .for_each(|(row, line)| {
            let line = line.unwrap();

            let mut current_numbers: Vec<(u32, usize, usize)> = Vec::new();
            let mut current_symbols: Vec<usize> = Vec::new();

            let mut current_start: Option<usize> = None;
            let mut current_value: u32 = 0;

            line.chars().enumerate().for_each(|(i, c)| {
                if c.is_numeric() {
                    if current_start.is_none() {
                        current_start = Some(i);
                    }
                    current_value = current_value * 10 + c.to_digit(10).unwrap();
                } else {
                    if let Some(j) = current_start {
                        current_numbers.push((current_value, j, i - 1));
                        current_start = None;
                        current_value = 0;
                    }

                    if c == '*' {
                        current_symbols.push(i);
                        gears.insert((row, i), Vec::new());
                    }
                }
            });

            if let Some(j) = current_start {
                current_numbers.push((current_value, j, line.len() - 1));
            }

            numbers.push(current_numbers);
            symbols.push(current_symbols);
        });

    numbers.into_iter().enumerate().for_each(|(i, number_row)| {
        for (value, start, end) in number_row {
            if i > 0 {
                for j in symbols[i - 1].clone() {
                    if (start as i32 - 1..=end as i32 + 1).contains(&(j as i32)) {
                        gears.get_mut(&(i - 1, j)).unwrap().push(value);
                    }
                }
            }
            if i < symbols.len() - 1 {
                for j in symbols[i + 1].clone() {
                    if (start as i32 - 1..=end as i32 + 1).contains(&(j as i32)) {
                        gears.get_mut(&(i + 1, j)).unwrap().push(value);
                    }
                }
            }
            for j in symbols[i].clone() {
                if (start as i32 - 1..=end as i32 + 1).contains(&(j as i32)) {
                    gears.get_mut(&(i, j)).unwrap().push(value);
                }
            }
        }
    });

    let mut total: u32 = 0;

    gears.values().filter(|x| x.len() == 2).for_each(|x| {
        total += x.iter().product::<u32>();
    });

    println!("{total}");
}

File: python/src/aoc2023/day3b.py
import sys


def run():
    symbols = {"*"}

    symbol_table: list[list[int]] = []  # row, column
    number_table: list[list[tuple[int, int, int]]] = []  # row, value, start, end
    gears: dict[tuple[int, int], list[int]] = {}

    for row, line in enumerate(sys.stdin):
        current_symbols: list[int] = []
        current_numbers: list[tuple[int, int, int]] = []

        current_start: None | int = None
        current_value = 0

        for i, c in enumerate(line):
            if c.isnumeric():
                if current_start is None:
                    current_start = i
                current_value = current_value * 10 + int(c)
            else:
                if current_start is not None:
                    current_numbers.append((current_value, current_start, i - 1))
                    current_start = None
                    current_value = 0
                if c in symbols:
                    current_symbols.append(i)
                    gears[(row, i)] = []

        if current_start is not None:
            current_numbers.append((current_value, current_start, len(line) - 1))

        symbol_table.append(current_symbols)
        number_table.append(current_numbers)

    for i, number_row in enumerate(number_table):
        for value, start, end in number_row:
            if i > 0:
                for j in symbol_table[i - 1]:
                    if start - 1 <= j <= end + 1:
                        gears[(i - 1, j)].append(value)
            if i < len(number_table) - 1:

                for j in symbol_table[i + 1]:
                    if start - 1 <= j <= end + 1:
                        gears[(i + 1, j)].append(value)

            for j in symbol_table[i]:
                if start - 1 <= j <= end + 1:
                    gears[(i, j)].append(value)

    valid_gears = [x for x in gears.values() if len(x) == 2]

    total = 0
    for a, b in valid_gears:
        total += a * b

    print(total)


if __name__ == "__main__":
    run()
